peps/pep-0793.rst

```diff
@@ -0,0 +1,476 @@
+temporarily switches to the main interpreter, calls the export hook
+there, and then either switches back and redoes the import, or fails.
+
+This unnecessary and fragile extra work highlights the underlying design issue:
+Python has no way to get information about an extension
+before the extension can, potentially, fully initialize itself.
+
+
+Rationale
+=========
+
+For avoiding the module export hook requiring a statically allocated
+``PyObject*``, two options come to mind:
+
+- Returning a *dynamically* allocated object, whose ownership is transferred
+  to the interpreter. This stucture could be very similar to the existing
+  ``PyModuleDef``, since it needs to contain the same data.
+  Unlike the existing ``PyModuleDef``, this one would need to be
+  reference-counted so that it both outlives “its” module and does not leak.
+
+- Adding a new export hook, which does not return a ``PyObject*``.
+
+  This was considered already for Python 3.5 in :pep:`489`, but rejected:
+
+    Keeping only the PyInit hook name, even if it’s not entirely appropriate
+    for exporting a definition, yielded a much simpler solution.
+
+  Alas, after a decade of fixing the implications of this choice, the solution
+  is no longer simple.
+
+A new hook will also allow Python to avoid the second issue mentioned in
+Motivation -- the interpreter switch.
+Effectivelly, it will add a new phase to multi-phase initialization, in which
+Python can check whether the module is compatible.
+
+
+Using slots without a wrapper struct
+------------------------------------
+
+The existing ``PyModuleDef`` is struct with some fixed fields and
+a “slots” array.
+Unlike slots, the fixed fields cannot be individually deprecated and replaced.
+This proposal does away with fixed fields and proposes using a slots array
+directly, without a wrapper struct.
+
+The ``PyModuleDef_Slot`` struct does have some downsides to fixed fields.
+We believe these are fixable, but leave that out of scope of this PEP
+(see “Improving slots in general” in the Possible Future Directions section).
+
+
+Tokens
+------
+
+A static ``PyModuleDef`` has another purpose besides describing
+how a module should be created.
+As a statically allocated singleton that remains attached to the module object,
+it allows extension authors to check whether a given Python module is “theirs”:
+if a module object has a known ``PyModuleDef``, its module state will have
+a known memory layout.
+
+An analogous issue was solved for types by adding ``Py_tp_token``.
+This proposal adds the same mechanism to modules.
+
+Unlike types, the import mechanism often has a pointer that's known to be
+suitable as a token value; in these cases it can provide a default token.
+Thus, module tokens do not need a varaint of the inelegant ``Py_TP_USE_SPEC``.
+
+
+Specification
+=============
+
+
+When importing an extension module, Python will newly look for an export hook
```

Comment:
Putting "newly" in this position in the sentence sounds weird. But it is clear what it means ... "now" would sound better.